### PR TITLE
Add BufferedImageChunk and GeometryModel

### PR DESCRIPTION
**Pull request by pyrollo**
Cette PR ajoute les classes nécessaires à la rasterisation des données vectorielles.

Une petite demo ajoutant des batiments sur la carte a été ajoutée au main. Elle peut se lancer ainsi:

```
cd minalac-generator
mvn compile && mvn exec:java -Dexec.mainClass="com.ignfab.minalac.generator.SampleImplementation" -Dexec.args="$HOME/.minetest/worlds/minalac EPSG:2154 600000 6340000 1000 1000 1.0 10.0"
```

### CODE REVIEW

**Comment by zhak5388** [comment]
Pas vraiment une revue, mais juste des commentaires en lisant le code (J'ai pas fini mais je publie quand-même)

File: src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     * Converts a geometry into wolrd coordinates.
+     *
+     * @param geom Geometry to transform.
+     * @return Transformed geometry (into world coordinates).
+     * @throws TransformException if unable to perform transformation.
+     */
+    public Geometry convert​(Geometry geom) throws TransformException {
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java]: Il y a un espace sans chasse qui a l'air de s'être glissé entre `convert` et `(` (Ca apparaît que ça soit en clonant ou bien en copiant depuis le repo)

File: src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+     * Construct a new {@code BufferedImageChunk}
+     *
+     * @param bbox Bounding box of that chunk
+     *
+     * Chunk is initalized with 0 values.
+     */
+    BufferedImageChunk(WorldBBox2d bbox) {
+        this.bbox = bbox;
+        this.image = new BufferedImage(bbox.getSize().getX(), bbox.getSize().getY(), BufferedImage.TYPE_BYTE_GRAY);
+        Arrays.fill(((DataBufferByte) image.getRaster().getDataBuffer()).getData(), (byte)0);
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java:39]: Je me demandais si il y avait une raison de prendre `BufferedImage.TYPE_BYTE_GRAY` spécifiquement (Enfin je me demandais s'il y avait un paramètre qui permettait d'initialiser à zéro directement)

Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java:39]: `TYPE_BYTE_GRAY` est censer utiliser un "int8". Le système de couleur complique tout ensuite mais cela semble le plus proche de ce que l'on veut (le reste, ça va être du RVB, donc 3 valeurs).

File: src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java (hunk 1 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java]: Corrigé. Le truc était presque invisible!

File: src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java (hunk 2 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java:39]: Documenter l'idée de cette bidouile avec les éventuelles pistes pour s'en sortir (modulo un bon développement)

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ * It is rasterizable. Rasterized chunk will include three values:
+ * 0: Not in geometry, 1: On its edge, 2: Inside it (polygons only).
+ */
+public class GeometryModel implements Rasterizable {
+    private Geometry geom;
+    private BufferedImageChunk chunk;
+
+    private static Color noneColor = BufferedImageChunk.colorFor(0); // Not in shape color
+    private static Color edgeColor = BufferedImageChunk.colorFor(1); // Shape border color
+    private static Color fillColor = BufferedImageChunk.colorFor(2); // Inside shape (fill) color
```
Comment (pyrollo) [inline src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Mettre des constantes ? ou des enums ?

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        // Until there is no need of it we don't keep original geometry.
+        // Geometry is stored transformed into world coordinates
+        this.geom = converter.convert(geom);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ReadableChunk2d getChunk() {
```
Comment (pyrollo) [inline src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Interable ? BufferedImageChunk ?

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                new WorldCoords2d((int)Math.floor(xMin), (int)Math.floor(yMin)),
+                new WorldCoords2d((int)Math.floor(xMax), (int)Math.floor(yMax)));
+
+            // Translate geometry into bounding box relative coordinates (i.e. BufferedImage coordinates)
+            Geometry geom = AffineTransformation.translationInstance(-Math.floor(xMin), -Math.floor(yMin)).transform(this.geom);
+
+            // Create chunk & draw geometry
+            chunk = new BufferedImageChunk(bbox);
+            Graphics2D graphics = chunk.createGraphics();
+            draw(graphics, geom);
```
Comment (pyrollo) [inline src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]:
```suggestion
            // Create chunk & draw geometry
            chunk = new BufferedImageChunk(bbox);
            Graphics2D graphics = chunk.createGraphics();
            draw(graphics, AffineTransformation.translationInstance(-Math.floor(xMin), -Math.floor(yMin)).transform(this.geom));
```

File: src/main/java/com/ignfab/minalac/generator/models/Rasterizable.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * Objects that can be rasterized into {@code ReadableChunk2d}.
+ */
+public interface Rasterizable {
+    /**
+     * Returns rasterized {@code ReadableChunk2d}.
+     *
+     * @return Rasterized chunk
+     */
+    ReadableChunk2d getChunk();
```
Comment (pyrollo) [inline src/main/java/com/ignfab/minalac/generator/models/Rasterizable.java:14]: Iterable ? à voir à l'usage.

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            double yMin = firstCoord.y;
+            double yMax = firstCoord.y;
+
+            for (Coordinate coord: geom.getCoordinates()) {
+                xMin = Math.min(xMin, coord.x);
+                xMax = Math.max(xMax, coord.x);
+                yMin = Math.min(yMin, coord.y);
+                yMax = Math.max(yMax, coord.y);
+            }
+
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Ça pourrait se simplifier je pense, il y a la méthode [`getEnvelopeInternal()`](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Geometry.html#getEnvelopeInternal--) qui renvoi le min et max de la BBOX. Ce qui donnerait:
```java
Envelope geomBBox = geom.getEnvelopeInternal();

double xMin = geomBBox.getMinX();
double xMax = geomBBox.getMaxX();
double yMin = geomBBox.getMinY();
double yMax = geomBBox.getMaxY();
```

File: src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    public void set(WorldCoords2d coords, int value) {
+        set(coords.getX(), coords.getY(), value);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Chunk2dIteratorSkip iterator() {
+        return new Chunk2dIteratorSkip((ReadableChunk2d)this, 0);
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java]: Je me demande si le cast est nécessaire `IterableChunk2d` est un `ReadableChunk2d` non?

File: src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+public class CoordsConverter {
+    private MathTransform crsTransform;
+    private AffineTransformation postTransform;
+
+    /**
+     * Constructs a new {@code CoordsConverter}.
+     *
+     * This is up to the caller to give proper transformations:
+     * @param crsTransform Transformation from source CRS to world CRS.
+     * @param postTransform Affine transformation applyed for scale and rotation after CRS transformation.
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java]: Petite erreur de typo qui s'est glissé : `applied` et non `applyed`

File: src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    public CoordsConverter(MathTransform crsTransform, AffineTransformation postTransform) {
+        if (crsTransform.getSourceDimensions() != 2 || crsTransform.getTargetDimensions() != 2)
+            throw new IllegalArgumentException("crsTransform must be from two dimensions to two dimensions");
+
+        this.crsTransform = crsTransform;
+        this.postTransform = postTransform;
+    }
+
+    /**
+     * Converts a geometry into wolrd coordinates.
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java]: Pareil typo : `world` et non `wolrd`

File: src/main/java/com/ignfab/minalac/generator/generation/Generation.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * Contains stuff about ongoing generation and its context.
+ */
+public class Generation {
+    private CoordinateReferenceSystem crs;
+    private AffineTransformation transformation;
+
+    /**
+     * Constructs a new generation context.
+     *
+     * @param crs Cordinate reference system used for generated world.
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/generation/Generation.java]: C'est `Coordinate` non?

File: src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+import com.ignfab.minalac.generator.utils.world2d.chunk.ReadableChunk2d;
+import com.ignfab.minalac.generator.utils.world2d.chunk.IterableChunk2d;
+import com.ignfab.minalac.generator.utils.world2d.chunk.WritableChunk2d;
+import com.ignfab.minalac.generator.utils.world2d.iterator.Chunk2dIteratorSkip;
+import com.ignfab.minalac.generator.utils.world2d.WorldCoords2d;
+import com.ignfab.minalac.generator.utils.world2d.WorldBBox2d;
+
+/**
+ * A readable anf writable 2d chunk of world map based on {@code BufferedImage}.
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java]: Typo: A readable *and* writable

File: src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    public Chunk2dIteratorSkip iterator() {
+        return new Chunk2dIteratorSkip((ReadableChunk2d)this, 0);
+    }
+
+    /**
+     * Creates a {@code Graphics2D} able to draw on this chunk.
+     *
+     * @return A {@code Graphics2D} instance linked to embedded {@code BufferedImage}.
+     *
+     * Coordinates in that {@code Graphics2D} are bounding box relative (starting at 0,0 and up to size).
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/models/BufferedImageChunk.java:95]: Il me semble que il est pas d'usage de mettre du texte après `@return`ça vient avant, mais je suis pas sur

File: src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
+package com.ignfab.minalac.generator.generation;
+
+import org.geotools.api.referencing.operation.MathTransform;
+import org.geotools.api.referencing.operation.TransformException;
+import org.geotools.geometry.jts.JTS;
+import org.geotools.referencing.CRS;
```
Comment (zhak5388) [inline src/main/java/com/ignfab/minalac/generator/generation/CoordsConverter.java]: C'est du détail mais je pense que l'import `CRS` n'est plus nécessaire (il doit peut être y avoir des imports non utilisés qui doivent trainer)

File: src/test/java/com/ignfab/minalac/models/TestBufferedImageChunk.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
+package com.ignfab.minalac.generator.models;
```
Comment (zhak5388) [inline src/test/java/com/ignfab/minalac/models/TestBufferedImageChunk.java]: Ici, je pense qu'il faut déplacer le package `models` dans le package `generation`. (Pour que ça suive en miroir main)

File: src/test/java/com/ignfab/minalac/generator/generation/TestCoordsConverter.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        assertEquals(-2.0, coords[0].getY(), 0.0001);
+        assertEquals(-2.0, coords[1].getX(), 0.0001);
+        assertEquals(-1.0, coords[1].getY(), 0.0001);
+        assertEquals(0.0, coords[2].getX(), 0.0001);
+        assertEquals(2.0, coords[2].getY(), 0.0001);
+    }
+
+    // Tests projection works as expected, combined or not with affine transformations
+    @Test
+    public void testProjection() throws TransformException, NoSuchAuthorityCodeException, FactoryException {
```
Comment (zhak5388) [inline src/test/java/com/ignfab/minalac/generator/generation/TestCoordsConverter.java:81]: C'est une idée mais je me demandais si ça pouvait être utile de faire une projection entre deux système non compatibles et s'assurer qu'il y a bien une erreur. (Par exemple entre [EPSG:29902](https://epsg.io/29902) et [EPSG:2154](https://epsg.io/2154))

**Review by zhak5388 — COMMENTED** [review]
J'ai mis quelques commentaires. C'est principalement sur de la typo + une broutille sur une méthode pour avoir min/max d'une `Geometry`

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java (hunk 3 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Fait

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java (hunk 4 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Fait ~~> BufferedImageChunk

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java (hunk 7 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java]: Fait !

File: src/test/java/com/ignfab/minalac/models/TestBufferedImageChunk.java (hunk 15 above)
Comment (pyrollo) [reply src/test/java/com/ignfab/minalac/models/TestBufferedImageChunk.java]: A discutter !

File: src/test/java/com/ignfab/minalac/generator/generation/TestCoordsConverter.java (hunk 16 above)
Comment (pyrollo) [reply src/test/java/com/ignfab/minalac/generator/generation/TestCoordsConverter.java:81]: En fait on testerai CRS.findMathTransform, ce qui n'est pas le sujet ici. En revanche dans la gestion des sources dans la génération, ça serait bon à tester.

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java [hunk 17]
```diff
@@ -6,50 +6,171 @@
...
 public class SampleImplementation {
-    public static void main(String[] args) throws IOException, OutOfWorldException, MapWriteException {
-        if (args.length != 5) {
-            System.out.println("There must be five arguments : directoryPath, baseURL, width, height, verticalScale");
+    public static void main(String[] args)
+    throws IOException, OutOfWorldException, MapWriteException, SAXException,
+           NoSuchAuthorityCodeException, FactoryException,
+           ParserConfigurationException, TransformException {
+        if (args.length != 8) {
+            System.out.println("There must be height arguments : directoryPath, crs, centerX, centerY, extendX, extendY, horizontalScale, verticalScale");
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/SampleImplementation.java]: Je veux bien mettre mes arguments en hauteur, mais ils sont 8 😄

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java [hunk 18]
```diff
@@ -6,50 +6,171 @@
...
+
+            String bboxURL = "BBOX=" + envelope.getMinX() + "," + envelope.getMinY() + "," + envelope.getMaxX() + "," + envelope.getMaxY();
+
+            // Creation of heightmap
+
+            System.out.println("Downloading height map");
+            HeightMap heightMap = createGroundHeightMap("https://data.geopf.fr/wms-r/wms?LAYERS=RGEALTI-MNT_PYR-ZIP_FXX_LAMB93_WMS&FORMAT=image/x-bil;bits=32&SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&STYLES=&CRS=" + crs + "&" + bboxURL, extendX, extendY, verticalScale);
+
+            // TODO: A lot of above code should end up in Generation class
+            Generation generation = new Generation(CRS.decode(crs), crsToVoxel); // This is supposed to be the target CRS
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:113]: Qu'est-ce que tu entends par `target CRS` ? C'est le pseudo CRS centré sur la map générée, et à l'échelle du bloc ?

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java [hunk 19]
```diff
@@ -6,50 +6,171 @@
...
+        }
+    }
+
+    private static void placeVectorFeatures(SimpleFeatureCollection features, CoordsConverter converter, VoxelWorld world, HeightMap heightMap)
+        throws TransformException, OutOfWorldException {
+        VoxelType insideVT = world.getFactory().createVoxelType(SemanticType.Cobble);
+        VoxelType wallVT = world.getFactory().createVoxelType(SemanticType.Brick);
+
+        SimpleFeatureIterator iterator = features.features();
+        try {
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/SampleImplementation.java]: Peut être simplifié avec un [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java [hunk 20]
```diff
@@ -6,50 +6,171 @@
...
+                SimpleFeature feature = iterator.next();
+
+                // Create a model out of feature geometry
+                GeometryModel model = new GeometryModel((Geometry)feature.getDefaultGeometry(), converter);
+
+                // Rasterize that model into a chunk
+                BufferedImageChunk chunk = model.getChunk();
+
+                // Iterate over chunk and draw shape on map at heightMap altitude
+                for (Chunk2dElement element : chunk) {
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:155]: C'est quoi un élément ? C'est un point (une zone d'1 bloc carré) qui appartient au chunk (à la géométrie ici) ?

File: src/main/java/com/ignfab/minalac/generator/inputs/WFS2DataProvider.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public SimpleFeatureCollection getFeatures() throws MalformedURLException, IOException, ParserConfigurationException, SAXException {
+        GML gml = new GML(GML.Version.GML3);
+
+        URL url = new URL(baseURL);
+
+        InputStream stream = url.openStream();
+
+        InputStream replacedStream =  new ReplacingInputStream( new ReplacingInputStream( new ReplacingInputStream( new ReplacingInputStream( stream,
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/inputs/WFS2DataProvider.java:31]: Un `ReplacingInputStream` ne peut remplacer qu'une seule chaîne...?

File: src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        // Until there is no need of it we don't keep original geometry.
+        // Geometry is stored transformed into world coordinates
+        this.geom = converter.convert(geom);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public BufferedImageChunk getChunk() {
```
Comment (indyteo) [inline src/main/java/com/ignfab/minalac/generator/models/GeometryModel.java:73]: Découper cette méthode en 2 (extraire la création du chunk) peut être utile pour la compéhension (et éventuellement si on veut forcer la création manuelle d'un chunk, pour les tests ou dans un edge case auquel on s'attendait pas)

File: src/test/java/com/ignfab/minalac/generator/models/TestBufferedImageChunk.java [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        WorldBBox2d bbox = new WorldBBox2d(-2, -1, 4, 3);
+        BufferedImageChunk chunk = new BufferedImageChunk(bbox);
+
+        try {
+            chunk.set(3, 0, 0);
+            fail("Exception expected when setting out of X bounds");
+        } catch (IndexOutOfBoundsException e) {
+        } catch (Exception e) {
+            fail("Wrong exception thrown when setting out of X bounds");
+        }
```
Comment (indyteo) [inline src/test/java/com/ignfab/minalac/generator/models/TestBufferedImageChunk.java]:
```suggestion
        assertThrows(IndexOutOfBoundsException.class, () -> chunk.set(3, 0, 0));
```

https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertThrows(java.lang.Class,org.junit.jupiter.api.function.Executable)

Par la même occasion, passer à JUnit 5 (aucune raison de rester sur la 4 qui a moins de features)

File: src/test/java/com/ignfab/minalac/generator/models/TestBufferedImageChunk.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        graphics.fillRect(1, 2, 3, 3);
+
+        for (int x = 2; x < 6; x++)
+            for (int y = 3; y < 8; y++)
+                if (x >= 3 && y >= 5)
+                    assertEquals("("+x+", "+y+")", 234, chunk.get(x, y));
+                else if (x <= 3 && y <= 5)
+                    assertEquals("("+x+", "+y+")", 123, chunk.get(x, y));
+                else
+                    assertEquals("("+x+", "+y+")", 0, chunk.get(x, y));
```
Comment (indyteo) [inline src/test/java/com/ignfab/minalac/generator/models/TestBufferedImageChunk.java]: Euuuuh, un petit commentaire ne ferait pas de mal je pense ? Ou peut-être plutôt une réécriture simplifiée...

Quand on se dit qu'il faut écrire un test unitaire pour tester le code de test, c'est pas bon signe 😁

**Review by indyteo — COMMENTED** [review]
Quelques commentaires après un premier coup d'œil

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java (hunk 17 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java]: ~~> Zen prépare une PR qui va drastiquement réduire le nb de paramètres :)

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java (hunk 18 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:113]: Alors il s'agit du CRS de la carte finale. C'est un vrai CRS (ce sera le 2154 pour la France par exemple). On applique ensuite dessus une translation, une rotation (plus tard) et une échelle.

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java (hunk 20 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:155]: Oui, c'est un élément placé (donc position + valeur), ou encore, le résultat d'un itérateur sur un chunk. La nomenclature est une vraie galère.

File: src/main/java/com/ignfab/minalac/generator/inputs/WFS2DataProvider.java (hunk 21 above)
Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/inputs/WFS2DataProvider.java:31]: Malheureusement oui.

**Review by indyteo — APPROVED** [review]
